### PR TITLE
fix(git-ignore): ignore symlinks correctly

### LINKED ISSUE #294: bug: --git-ignore not working with globbings on symlinks on ubuntu
eza version: v0.11.1 [+git]
command: `eza -lah --git --git-ignore`
.gitignore file affected content:
```
bazel-*
```

Output without --git-ignore:
```
.rw-rw-r--    23 mfn  29 août  10:00  -- .bazelignore
.rw-rw-r--   718 mfn  29 août  10:00  -- .bazelrc
drwxrwxr-x     - mfn  12 sept. 16:56  -I .git
.rw-rw-r--    26 mfn  29 août  10:00  -- .gitattributes
drwxrwxr-x     - mfn  29 août  10:00  -- .github
.rw-rw-r--   284 mfn  29 août  10:00  -- .gitignore
.rw-rw-r--   686 mfn  29 août  10:00  -- .gitlint
drwxrwxr-x     - mfn  29 août  10:00  -- .http
drwxrwxr-x     - mfn  29 août  15:47  -I .ijwb
drwxrwxr-x     - mfn  31 août  11:28  -I .metals
drwxrwxr-x     - mfn  29 août  10:00  -- .run
.rw-rw-r--   577 mfn  29 août  10:00  -- .scalafmt.conf
drwxrwxr-x     - mfn  31 août  11:38  -I .trunk
drwxrwxr-x     - mfn  29 août  16:28  -I .vscode
drwxrwxr-x     - mfn   5 sept. 15:51  -- 3rdparty
drwxrwxr-x     - mfn  29 août  10:00  -- adex-services
drwxrwxr-x     - mfn  29 août  10:00  -- api
drwxrwxr-x     - mfn  29 août  10:00  -- apps
drwxrwxr-x     - mfn  29 août  10:00  -- bank-accounting-services
lrwxrwxrwx     - mfn  11 sept. 16:21  -- bazel-bin -> /home/mfn/.cache/bazel/_bazel_mfn/4c4fc07d94d1d7de773541c26cf1677e/execroot/mediarithmics_platform/bazel-out/k8-fastbuild/bin
lrwxrwxrwx     - mfn  11 sept. 16:21  -- bazel-mediarithmics-platform -> /home/mfn/.cache/bazel/_bazel_mfn/4c4fc07d94d1d7de773541c26cf1677e/execroot/mediarithmics_platform
lrwxrwxrwx     - mfn  11 sept. 16:21  -- bazel-out -> /home/mfn/.cache/bazel/_bazel_mfn/4c4fc07d94d1d7de773541c26cf1677e/execroot/mediarithmics_platform/bazel-out
lrwxrwxrwx     - mfn  11 sept. 16:21  -- bazel-testlogs -> /home/mfn/.cache/bazel/_bazel_mfn/4c4fc07d94d1d7de773541c26cf1677e/execroot/mediarithmics_platform/bazel-out/k8-fastbuild/testlogs
.rw-rw-r--   685 mfn  12 sept. 16:57  -I bazel_explain.log
drwxrwxr-x     - mfn  12 sept. 15:03  -- build-support
.rw-rw-r--  3,0k mfn  29 août  10:00  -- BUILD.bazel
drwxrwxr-x     - mfn  29 août  10:00  -- business-analytics
drwxrwxr-x     - mfn  29 août  10:00  -- campaign-delivery
drwxrwxr-x     - mfn  29 août  10:00  -- core-platform-services
drwxrwxr-x     - mfn  29 août  10:00  -- cql-schema
drwxrwxr-x     - mfn  29 août  10:00  -- datamart
.rw-rw-r--   19k mfn   5 sept. 15:51  -- dependencies.yaml
drwxrwxr-x     - mfn  29 août  10:00  -- deployment
.rw-rw-r--  4,8k mfn  29 août  10:00  -- deployment.md
drwxrwxr-x     - mfn  29 août  10:00  -- docker-example
drwxrwxr-x     - mfn  29 août  10:00  -- docs
drwxrwxr-x     - mfn  29 août  10:00  -- graphdb
drwxrwxr-x     - mfn  29 août  10:00  -- infra
drwxrwxr-x     - mfn  29 août  10:00  -- job-management-services
drwxrwxr-x     - mfn  29 août  10:00  -- lib
drwxrwxr-x     - mfn  29 août  10:00  -- mocks
drwxrwxr-x     - mfn  29 août  10:00  -- plugin-services
.rw-rw-r--  1,5k mfn  29 août  10:00  -- README.md
.rw-rw-r--  2,9k mfn  29 août  10:00  -- sonar-project.properties
drwxrwxr-x     - mfn  12 sept. 15:03  -- src
drwxrwxr-x     - mfn  12 sept. 15:03  -- tools
drwxrwxr-x     - mfn  29 août  10:00  -- types
drwxrwxr-x     - mfn  29 août  10:00  -- user-mapping-services
drwxrwxr-x     - mfn  29 août  10:00  -- virtual-platform
.rw-rw-r--  6,0k mfn   5 sept. 15:51  -- WORKSPACE
```
Output with --git-ignore:
```
.rw-rw-r--    23 mfn  29 août  10:00  -- .bazelignore
.rw-rw-r--   718 mfn  29 août  10:00  -- .bazelrc
.rw-rw-r--    26 mfn  29 août  10:00  -- .gitattributes
drwxrwxr-x     - mfn  29 août  10:00  -- .github
.rw-rw-r--   284 mfn  29 août  10:00  -- .gitignore
.rw-rw-r--   686 mfn  29 août  10:00  -- .gitlint
drwxrwxr-x     - mfn  29 août  10:00  -- .http
drwxrwxr-x     - mfn  29 août  10:00  -- .run
.rw-rw-r--   577 mfn  29 août  10:00  -- .scalafmt.conf
drwxrwxr-x     - mfn   5 sept. 15:51  -- 3rdparty
drwxrwxr-x     - mfn  29 août  10:00  -- adex-services
drwxrwxr-x     - mfn  29 août  10:00  -- api
drwxrwxr-x     - mfn  29 août  10:00  -- apps
drwxrwxr-x     - mfn  29 août  10:00  -- bank-accounting-services
lrwxrwxrwx     - mfn  11 sept. 16:21  -- bazel-bin -> /home/mfn/.cache/bazel/_bazel_mfn/4c4fc07d94d1d7de773541c26cf1677e/execroot/mediarithmics_platform/bazel-out/k8-fastbuild/bin
lrwxrwxrwx     - mfn  11 sept. 16:21  -- bazel-mediarithmics-platform -> /home/mfn/.cache/bazel/_bazel_mfn/4c4fc07d94d1d7de773541c26cf1677e/execroot/mediarithmics_platform
lrwxrwxrwx     - mfn  11 sept. 16:21  -- bazel-out -> /home/mfn/.cache/bazel/_bazel_mfn/4c4fc07d94d1d7de773541c26cf1677e/execroot/mediarithmics_platform/bazel-out
lrwxrwxrwx     - mfn  11 sept. 16:21  -- bazel-testlogs -> /home/mfn/.cache/bazel/_bazel_mfn/4c4fc07d94d1d7de773541c26cf1677e/execroot/mediarithmics_platform/bazel-out/k8-fastbuild/testlogs
drwxrwxr-x     - mfn  12 sept. 15:03  -- build-support
.rw-rw-r--  3,0k mfn  29 août  10:00  -- BUILD.bazel
drwxrwxr-x     - mfn  29 août  10:00  -- business-analytics
drwxrwxr-x     - mfn  29 août  10:00  -- campaign-delivery
drwxrwxr-x     - mfn  29 août  10:00  -- core-platform-services
drwxrwxr-x     - mfn  29 août  10:00  -- cql-schema
drwxrwxr-x     - mfn  29 août  10:00  -- datamart
.rw-rw-r--   19k mfn   5 sept. 15:51  -- dependencies.yaml
drwxrwxr-x     - mfn  29 août  10:00  -- deployment
.rw-rw-r--  4,8k mfn  29 août  10:00  -- deployment.md
drwxrwxr-x     - mfn  29 août  10:00  -- docker-example
drwxrwxr-x     - mfn  29 août  10:00  -- docs
drwxrwxr-x     - mfn  29 août  10:00  -- graphdb
drwxrwxr-x     - mfn  29 août  10:00  -- infra
drwxrwxr-x     - mfn  29 août  10:00  -- job-management-services
drwxrwxr-x     - mfn  29 août  10:00  -- lib
drwxrwxr-x     - mfn  29 août  10:00  -- mocks
drwxrwxr-x     - mfn  29 août  10:00  -- plugin-services
.rw-rw-r--  1,5k mfn  29 août  10:00  -- README.md
.rw-rw-r--  2,9k mfn  29 août  10:00  -- sonar-project.properties
drwxrwxr-x     - mfn  12 sept. 15:03  -- src
drwxrwxr-x     - mfn  12 sept. 15:03  -- tools
drwxrwxr-x     - mfn  29 août  10:00  -- types
drwxrwxr-x     - mfn  29 août  10:00  -- user-mapping-services
drwxrwxr-x     - mfn  29 août  10:00  -- virtual-platform
.rw-rw-r--  6,0k mfn   5 sept. 15:51  -- WORKSPACE
```

ArchLinux Output:
```
.rw-rw-r--    23 mfn  29 août  10:00  -- .bazelignore
.rw-rw-r--   718 mfn  29 août  10:00  -- .bazelrc
.rw-rw-r--    26 mfn  29 août  10:00  -- .gitattributes
drwxrwxr-x     - mfn  29 août  10:00  -- .github
.rw-rw-r--   284 mfn  29 août  10:00  -- .gitignore
.rw-rw-r--   686 mfn  29 août  10:00  -- .gitlint
drwxrwxr-x     - mfn  29 août  10:00  -- .http
drwxrwxr-x     - mfn  29 août  10:00  -- .run
.rw-rw-r--   577 mfn  29 août  10:00  -- .scalafmt.conf
drwxrwxr-x     - mfn   5 sept. 15:51  -- 3rdparty
drwxrwxr-x     - mfn  29 août  10:00  -- adex-services
drwxrwxr-x     - mfn  29 août  10:00  -- api
drwxrwxr-x     - mfn  29 août  10:00  -- apps
drwxrwxr-x     - mfn  29 août  10:00  -- bank-accounting-services
drwxrwxr-x     - mfn  12 sept. 15:03  -- build-support
.rw-rw-r--  3,0k mfn  29 août  10:00  -- BUILD.bazel
drwxrwxr-x     - mfn  29 août  10:00  -- business-analytics
drwxrwxr-x     - mfn  29 août  10:00  -- campaign-delivery
drwxrwxr-x     - mfn  29 août  10:00  -- core-platform-services
drwxrwxr-x     - mfn  29 août  10:00  -- cql-schema
drwxrwxr-x     - mfn  29 août  10:00  -- datamart
.rw-rw-r--   19k mfn   5 sept. 15:51  -- dependencies.yaml
drwxrwxr-x     - mfn  29 août  10:00  -- deployment
.rw-rw-r--  4,8k mfn  29 août  10:00  -- deployment.md
drwxrwxr-x     - mfn  29 août  10:00  -- docker-example
drwxrwxr-x     - mfn  29 août  10:00  -- docs
drwxrwxr-x     - mfn  29 août  10:00  -- graphdb
drwxrwxr-x     - mfn  29 août  10:00  -- infra
drwxrwxr-x     - mfn  29 août  10:00  -- job-management-services
drwxrwxr-x     - mfn  29 août  10:00  -- lib
drwxrwxr-x     - mfn  29 août  10:00  -- mocks
drwxrwxr-x     - mfn  29 août  10:00  -- plugin-services
.rw-rw-r--  1,5k mfn  29 août  10:00  -- README.md
.rw-rw-r--  2,9k mfn  29 août  10:00  -- sonar-project.properties
drwxrwxr-x     - mfn  12 sept. 15:03  -- src
drwxrwxr-x     - mfn  12 sept. 15:03  -- tools
drwxrwxr-x     - mfn  29 août  10:00  -- types
drwxrwxr-x     - mfn  29 août  10:00  -- user-mapping-services
drwxrwxr-x     - mfn  29 août  10:00  -- virtual-platform
.rw-rw-r--  6,0k mfn   5 sept. 15:51  -- WORKSPACE
```

This looks like a ubuntu specific error as I also uses eza on arch and it works, never previously worked on exa ubuntu.

### PR DESCRIPTION
**Pull request by ariasuni**
Port of https://github.com/ogham/exa/pull/872 to eza.

The test is in the exa PR is passing fine, and I’m not sure if there are edge cases I didn’t think of.

Edit:
fix #294

### CODE REVIEW

**Comment by cafkafk** [comment]
We're in the middle of writing a lot of integration tests, so it would probably be better to merge this after.

File: src/fs/feature/git.rs [hunk 1]
```diff
@@ -319,15 +319,10 @@ impl Git {
...
 fn reorient(path: &Path) -> PathBuf {
-    use std::env::current_dir;
-
-    // TODO: I’m not 100% on this func tbh
-    let path = match current_dir() {
-        Err(_)   => Path::new(".").join(path),
-        Ok(dir)  => dir.join(path),
+    lazy_static! {
+        static ref CWD: PathBuf = std::env::current_dir().unwrap();
     };
```
Comment (cafkafk) [inline src/fs/feature/git.rs:324]: We have rust `1.70` now, maybe this could be a `OnceLock`?

**Review by cafkafk — CHANGES_REQUESTED** [review]

**Comment by ariasuni** [comment]
This code is still wrong, the `canonicalize()` is what is causing #294. I think just using `CWD.with_file_name(path)` should work, and that we should also use that instead of `canonicalize()` for hyperlinks since it’s much faster (doesn’t access the file system etc.). So, changing this to draft for now.

File: src/fs/file.rs [hunk 2]
```diff
@@ -103,7 +103,7 @@ impl<'dir> File<'dir> {
         let extended_attributes = OnceLock::new();
         let absolute_path = OnceLock::new();
```
Comment (cfxegbert) [inline src/fs/file.rs:104]: Rename this to `canonical_path` and you can avoid the `canonical_path: absolute_path` below.

File: src/fs/file.rs [hunk 3]
```diff
@@ -117,7 +117,7 @@ impl<'dir> File<'dir> {
...
```
Comment (cfxegbert) [inline src/fs/file.rs:118]: Same as above

File: src/fs/file.rs [hunk 4]
```diff
@@ -130,7 +130,7 @@ impl<'dir> File<'dir> {
...
```
Comment (cfxegbert) [inline src/fs/file.rs:131]: Same as above

File: src/fs/file.rs [hunk 5]
```diff
@@ -259,44 +259,32 @@ impl<'dir> File<'dir> {
         self.metadata.file_type().is_socket()
     }
 
-    /// Determine the full path resolving all symbolic links on demand.
-    pub fn absolute_path(&self) -> Option<&PathBuf> {
-        self.absolute_path.get_or_init(|| std::fs::canonicalize(&self.path).ok()).as_ref()
+    /// Determine the absolute path without resolving symlinks.
+    pub fn absolute_path(&self) -> PathBuf {
+        static CWD: OnceLock<PathBuf> = OnceLock::new();
+        CWD.get_or_init(|| std::env::current_dir().unwrap()).with_file_name(&self.path)
```
Comment (cfxegbert) [inline src/fs/file.rs:265]: This is different than how canonicalize works above.  I'll have to think about the difference for a bit.  Will this work if the user enters `eza -l --mount ../../foo`?

File: src/fs/file.rs [hunk 6]
```diff
@@ -319,7 +307,7 @@ impl<'dir> File<'dir> {
             Err(e)  => return FileTarget::Err(e),
         };
 
-        let absolute_path = self.reorient_target_path(&path);
+        let absolute_path = self.path.with_file_name(&path);
```
Comment (cfxegbert) [inline src/fs/file.rs:310]: Same renaming comment

File: src/fs/file.rs (hunk 5 above)
Comment (ariasuni) [reply src/fs/file.rs:265]: Indeed, now that I look at my code with fresh eyes, this look… more than dubious. It won’t work if `self.path` is in a subdirectory of the current directory, or when using `..` and `.`.

I need to use [`Path::join()`](https://doc.rust-lang.org/std/path/struct.Path.html#method.join) instead, it’ll normalize the path and it should work in all cases.

**Comment by ariasuni** [comment]
Turns out this is a very annoying problem. `libgit2` gives us absolute, canonicalized paths. So, imagine an hypothetical Git repository:

```
/
├── repository
│   ├── src/
│   ├── link/ -> ../target/
│   └── Cargo.toml
├── repository_symlink -> ./repository
└── target
    └── file
```

So now imagine someone is running `eza --tree --git` from `/repository_symlink/`, we would like to get the absolute path with symlink resolved `/repository/` to be able to compare it to the output of the `libgit2`, but when we get to `/repository_symlink/link/target/file`, we’d likely want to get the intermediate form (one symlink resolved, one symlink not resolved) `/repository/link/target/file` so that we can see if it’s in the repository.

Everything starts being even more complicated when you take into account that you can pass several paths to the command line. I suppose the logical thing would to take the initial path passed to the command line, canonicalize it, then adjust the files to this path, which means passing around the original_path which is annoying. But then, we could just canonicalize the paths from the command line, because it would relatively cheap (passing a lot of argument to the command line should be relatively rare).

But also now I realize that the code uses the workdir (absolute, canonicalized path) to get the absolute_path of the files in Git, so that we should canonicalize paths before trying to get their status, which may not work so well some of the path (see paragraph about `--tree` and symlinks), so I have to rethink everything properly so that paths and symlinks and git stop being such a mess :sob: probably by creating a new `GitRepoView` that uses the path we’re traversing at the moment we’re trying to get the Git status, that should be doable, but it needs some work.

Also it would probably be good to write tests afterwards, I’m not sure if we can create a Git repository in integration tests?

**Comment by cafkafk** [comment]
> I’m not sure if we can create a Git repository in integration tests?

We actually can! In the justfile, I actually even think I already did something with git, we just don't use the actual test dir generation yet... but we will!

**Comment by ariasuni** [comment]
I think most of this is irrelevant now, and I opened https://github.com/eza-community/eza/pull/1436 for the rest.
